scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```
@@ -490,20 +521,65 @@ object Execution {
-      cache.getOrElseInsert(this,
-        for {
-          flowDef <- toFuture(Try { val fd = new FlowDef; (head :: tail).foreach(_.write(conf, fd, mode)); fd })
-          _ = FlowStateMap.validateSources(flowDef, mode)
-          jobStats <- cache.runFlowDef(conf, mode, flowDef)
-          _ = FlowStateMap.clear(flowDef)
-        } yield (fn(conf, mode), ExecutionCounters.fromJobStats(jobStats)))
+
+    /* Run a list of ToWrite elements */
+    private[this] def scheduleToWrites(conf: Config,
+      mode: Mode,
+      cache: EvalCache,
+      head: ToWrite,
+      tail: List[ToWrite])(implicit cec: ConcurrentExecutionContext): Future[ExecutionCounters] = {
+      for {
+        flowDef <- toFuture(Try { val fd = new FlowDef; (head :: tail).foreach(_.write(conf, fd, mode)); fd })
+        _ = FlowStateMap.validateSources(flowDef, mode)
+        jobStats <- cache.runFlowDef(conf, mode, flowDef)
+        _ = FlowStateMap.clear(flowDef)
+      } yield (ExecutionCounters.fromJobStats(jobStats))
+    }
+
+    def unwrapListEither[A, B, C](it: List[(A, Either[B, C])]): (List[(A, B)], List[(A, C)]) = it match {
+      case (a, Left(b)) :: tail =>
+        val (l, r) = unwrapListEither(tail)
+        ((a, b) :: l, r)
+      case (a, Right(c)) :: tail =>
+        val (l, r) = unwrapListEither(tail)
+        (l, (a, c) :: r)
+      case Nil => (Nil, Nil)
+    }
+
+    // We look up to see if any of our ToWrite elements have already been ran
+    // if so we remove them from the cache.
+    // Anything not already ran we run as part of a single flow def, using their combined counters for the others
+    def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) = {
+      val cacheLookup: List[(ToWrite, Either[Promise[ExecutionCounters], Future[ExecutionCounters]])] = (head :: tail).map{ tw => (tw, cache.getOrLock(tw)) }
+      val (weDoOperation, someoneElseDoesOperation) = unwrapListEither(cacheLookup)
+
```

Thats a good idea, can just fail fast the whole process